Default Token Address Linking

wilwade opened the pull request:
## Goal

The goal of this PR is:
- Introduce a new schema for MSAs to be able to attach wallet addresses for payment to an MSA
- Setup the future structure for Frequency schemas based on https://github.com/LibertyDSNP/schemas

Closes #3

## Discussion

- *Question*: Itemized or Paginated? `Current Winner: Itemized`
    - Itemized: Requires a bit more work to keep up to date
    - Paginated: Easy editing of multiple and removal of out of date items
- *Question*: Indexed by `coin_type` or Array of Objects (Itemized Required)? `Current Winner: Indexed`
    - Indexed: Requires an array structure after it still to support multiple addresses with the same coin_type
    - Array: Has to be scanned fully each time, but likely a small dataset for most users

## Checklist

- [x] PR Self-Review and Commenting
- [x] Docs updated
- [x] Tests added

## How To Test the Changes

1. Clone the pr branch
2. `npm run test`
3. Start Frequency Chain
4. `npm run deploy`

saraswatpuneet commented on schemas/index.js:
```diff
@@ -0,0 +1,13 @@
+      modelType: "AvroBinary",
+      payloadLocation: "Itemized",
+      settings: ["SignatureRequired"],
+    },
```
Do we need `AppendOnly` for itemized schema ?

wilwade replied:
In this case it is OK for items to be removed without issue.

wesbiggs commented on schemas/walletAddresses.js:
```diff
@@ -0,0 +1,22 @@
+ */
+export default {
+  type: "record",
+  name: "Address",
```
This seems a little redundant with the "address" field; maybe WalletAddress for the name?

wilwade replied:
I was back and forth. I don't know if WalletAddress was better or if we needed something like ChainAddress (as they all have to be chain addresses?) That said the schema is `WalletAddresses` so I guess the atom would be `WalletAddress`

Open to ideas.

wilwade replied:
Updated to Payment Address with the other payment address changes

wesbiggs commented on schemas/walletAddresses.js:
```diff
@@ -0,0 +1,22 @@
+  namespace: "xyz.frequency",
+  fields: [
+    {
+      name: "coin_type",
```
I see that SLIP-0044 calls them coin types, but should we consider the slightly more neutral "token_type" instead?

wilwade replied:
I went back and forth on this. I am tempted toward the even more specific `slip_0044_coin_type` so that the documentation is even clearer what sort of data it is, but that's a bit much perhaps.

I want to keep it as connected as possible to the source for this data as it is otherwise only documentation that tells you how to read the data.

Perhaps I should switch to `slip_0044_coin_type` or just `slip_0044`?

enddynayn replied:
I vote slip_0044_coin_type

wilwade replied:
Went for `coin_type_slip_0044` (so coin type is still first and easier to remember with code completion)

wesbiggs reviewed (commented):
Looks good -- I just left a couple of naming questions (that you may have thought about already). I am still working through how this repo relates to the (forthcoming) DSNP schemas repo, but this all makes sense for walletAddresses.

JoeCap08055 commented:
Recommend we reference an issue with this PR, that describes the problem being solved & other points, such as why we can't simply use 'addPublicKeyToMsa' (presumably because they're not supported key types, ie sr15519). Could also enhance the "Goal" in the README with the same information.

wilwade commented:
> Recommend we reference an issue with this PR, that describes the problem being solved & other points, such as why we can't simply use 'addPublicKeyToMsa' (presumably because they're not supported key types, ie sr15519). Could also enhance the "Goal" in the README with the same information.

@JoeCap08055 Good points.

- Added issue #2  (although didn't specifically address addPublicKeyToMsa, I think it is more clear)
- Updated the README a bunch

enddynayn reviewed (approved):
👍 awesome!

saraswatpuneet reviewed (approved):
Lgtm

JoeCap08055 commented:
Not trying to be picky, but I wonder if there's a better term than "payment address" that would be more clear that these are NOT (necessarily) FRQCY-holding wallets for paying transaction fees, but are simply for informational purposes

wilwade commented:
> Not trying to be picky, but I wonder if there's a better term than "payment address" that would be more clear that these are NOT (necessarily) FRQCY-holding wallets for paying transaction fees, but are simply for informational purposes

@JoeCap08055 I am 100% open to options. "Token Addresses" didn't feel right, but perhaps? 

Some additional meh ideas:
- Payment Target
- Token Wallet
- Transfer Address
- Token Linked Address

JoeCap08055 commented:
@wilwade I'm similarly stumped as to a good name; here are some ChatGPT-generated options, maybe one will provide some inspiration:

1. **Payment Address Registry**: Emphasizes that the addresses are for receiving payments across different tokens and blockchains.

2. **Cross-Chain Payment List**: Highlights the chain-agnostic nature of the addresses.

3. **Token-Agnostic Address Book**: Focuses on the token-agnostic aspect of the list.

4. **Receiver Address Pool**: Suggests that the addresses are meant for receiving payments rather than sending.

5. **Payment Destination Directory**: Conveys that the addresses are destinations for payments, independent of the blockchain.

6. **Universal Payment Address List**: Emphasizes the universal applicability of the addresses.

7. **External Payment Address List**: Indicates that these addresses are external to the Frequency blockchain in terms of payment capability.

8. **Token Receiver Address Book**: Focuses on the role of these addresses as receivers of various tokens.

9. **Interchain Payment Address List**: Highlights the interoperability across different blockchains.

10. **Wallet Destination Registry**: Indicates that the addresses are associated with a wallet for receiving tokens.
